### PR TITLE
feat(launch): wash forensics — Helius wallet-concentration evidence (#209)

### PR DESCRIPTION
**Pull request by irfndi**
Wash forensics for the launch lane — distinguish real volume from wash trading before capital enters.

## Grounded in live data
Verified the Helius enhanced API against the real TOAD pool: its recent 40 txs came from **4 feePayers, 25 from one wallet, in a 4-second window**. DLMM swaps aren't in Helius's parsed models (type UNKNOWN, no tokenTransfers) — but the **feePayer per tx survives**, and that's all the signals need.

## What it does
- **Pure score** (`engine/wash-forensics.ts`, tested): wallet concentration (≤2 payers across ≥20 trades, or ≤15% distinct-payer rate), bot bursts (>5 trades/sec from ≤2 wallets), advisory fee uniformity. **Hard-only at the extreme tail** — a real launch's early volume is naturally concentrated.
- **One Helius call per admitted pool per radar refresh** (`/v0/addresses/{pool}/transactions?limit=40`), cached per pool, consumed by the ENTER gate.
- **`LAUNCH_WASH_FORENSICS_ENABLED`** (default false): the radar log carries `wash: { suspicious, reason, distinctPayers, tradeCount }` per admitted pool; an audited `[wash-forensics]` ENTER gate rejects egregious evidence before capital enters a honeypot's volume.
- **Fail-open everywhere**: fetch/parse errors, a non-Helius RPC host, or the switch off → null → never blocks.

## Tests
8 score cases: the live TOAD shape, 2-wallet concentration, bot burst, organic volume, thin-sample floor, lone-wallet tradeoff, uniform-fee advisory, empty. 1681/1681; tsc 0; oxlint 0.

## Summary by Sourcery

Introduce wash-trading forensics for launch pools, scoring recent Helius transaction data to optionally block egregious wash volume before capital ENTER.

New Features:
- Add a wash-forensics scoring module that analyzes wallet concentration, burst density, and fee uniformity from recent pool transactions.
- Integrate an optional wash evidence fetch via Helius enhanced API for admitted launch pools, exposing findings in the launch radar log.
- Gate launch ENTER decisions on strong wash-trading evidence when the wash forensics feature flag is enabled, rejecting suspicious pools before entry.

Enhancements:
- Extend application configuration with a LAUNCH_WASH_FORENSICS_ENABLED switch, defaulting to advisory-only logging when disabled.
- Augment launch radar logging with per-pool wash evidence metadata (suspicious flag, reason, distinct payers, trade count) when available.

Tests:
- Add unit tests covering multiple wash evidence scenarios, including concentrated samples, bot bursts, organic volume, thin samples, lone-wallet patterns, fee uniformity, and empty inputs.

## Summary by CodeRabbit

* **New Features**
  * Added optional wash-trading forensics for launch pools.
  * Pools with suspicious trading patterns are now blocked from launch entry, while missing or unavailable evidence does not prevent entry.
  * Added detection for concentrated wallets, low payer diversity, and high-volume bursts.
  * Added configuration to enable or disable wash-trading checks.

* **Tests**
  * Added coverage for concentrated activity, bot bursts, organic trading, sparse data, fee patterns, and empty inputs.

### CODE REVIEW

File: engine/adapter-service.ts [hunk 1]
```diff
@@ -2104,6 +2105,52 @@ export const AdapterLive = Layer.effect(
...
+        Effect.gen(function* () {
+          if (config.launchWashForensicsEnabled !== true) return null;
+          if (!config.heliusApiKey) return null;
+          let host: string;
+          try {
+            host = new URL(config.solanaRpcUrl).host;
+          } catch {
+            return null;
+          }
+          if (!host.includes("helius")) return null;
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/adapter-service.ts]: Fixed: the Helius hostname is now validated as an exact domain (helius-rpc.com or *.helius-rpc.com) — a custom host merely containing the substring could no longer receive the API key.

File: engine/program.ts [hunk 2]
```diff
@@ -4211,23 +4216,44 @@ export const program = Effect.gen(function* () {
           launchScanPools.add(r.pool.address);
         }
       }
+      if (config.launchWashForensicsEnabled === true && adapter.getPoolWashEvidence) {
+        for (const r of gateResult.ranked.slice(0, config.launchScanTopK ?? 30)) {
+          const evidence = yield* adapter
+            .getPoolWashEvidence(r.pool.address)
+            .pipe(Effect.catch(() => Effect.succeed(null)));
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/program.ts]: Fixed: the evidence fetches run in parallel (Effect.forEach concurrency: unbounded) with a 5s timeout — an enabled refresh is bounded by the slowest single call, not the sum (was up to 5 min worst case at 30 x 10s).

File: engine/program.ts [hunk 3]
```diff
@@ -4211,23 +4216,44 @@ export const program = Effect.gen(function* () {
...
+          if (evidence !== null) washEvidenceByPool.set(r.pool.address, evidence);
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/program.ts]: Fixed: washEvidenceByPool is cleared before every refresh and only re-populated on non-null results — a fetch failure (null) or a pool leaving the top-K drops its entry instead of leaving an old suspicious flag gating ENTER indefinitely.

File: engine/config-service.ts [hunk 4]
```diff
@@ -242,6 +242,8 @@ export interface AppConfig {
   readonly launchRunnerScaleInStepPct?: number;
   readonly launchRunnerScaleInSizePct?: number;
   readonly launchRunnerScaleInMaxSteps?: number;
+  /** Wash forensics master switch. */
+  readonly launchWashForensicsEnabled?: boolean;
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/config-service.ts:246]: Declined with the codebase convention: launchWashForensicsEnabled stays optional alongside the 15+ sibling launch knobs, and getPoolWashEvidence stays an optional adapter method like discoverHotPools/discoverPoolsTopPages — the live layer always provides it and the switch gates the cost. Making only this one required would be inconsistent.

File: engine/program.ts [hunk 5]
```diff
@@ -4211,23 +4216,44 @@ export const program = Effect.gen(function* () {
...
+        }
+      }
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/program.ts:4243]: Fixed: the map is cleared before each refresh and re-populated only from non-null results — stale suspicious flags cannot outlive their pool or a fetch failure.

File: engine/program.ts [hunk 6]
```diff
@@ -4211,23 +4216,44 @@ export const program = Effect.gen(function* () {
           launchScanPools.add(r.pool.address);
         }
       }
+      if (config.launchWashForensicsEnabled === true && adapter.getPoolWashEvidence) {
+        for (const r of gateResult.ranked.slice(0, config.launchScanTopK ?? 30)) {
+          const evidence = yield* adapter
+            .getPoolWashEvidence(r.pool.address)
+            .pipe(Effect.catch(() => Effect.succeed(null)));
+          if (evidence !== null) washEvidenceByPool.set(r.pool.address, evidence);
+        }
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/program.ts]: Fixed: parallel fetch (concurrency: unbounded) with the 5s per-call timeout; also see the hostname + staleness fixes in the same round.

File: engine/program.ts [hunk 7]
```diff
@@ -4211,23 +4216,44 @@ export const program = Effect.gen(function* () {
...
+                : Math.max(0, (now - r.pool.createdAtMs) / 3_600_000),
+            ...(evidence !== undefined
+              ? {
+                  wash: {
+                    suspicious: evidence.suspicious,
+                    reason: evidence.reason,
+                    distinctPayers: evidence.distinctPayers,
+                    tradeCount: evidence.tradeCount,
+                  },
+                }
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/program.ts:4273]: Fixed: the radar payload now carries the full evidence — uniquePayerRate, txsPerSecond, feeCv — so fee uniformity is auditable, not just computed.

File: engine/wash-forensics.ts [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+/** > this many trades/sec from ≤2 wallets = bot burst. */
+export const WASH_MAX_TPS_FOR_BURST = 5;
+
+export function scoreWashEvidence(rows: ReadonlyArray<WashTradeRow>): WashEvidence {
+  const tradeCount = rows.length;
+  const distinctPayers = new Set(rows.map((r) => r.payer)).size;
+  const timestamps = rows.map((r) => r.timestamp);
+  const spanSec =
+    tradeCount >= 2 ? Math.max(1, Math.max(...timestamps) - Math.min(...timestamps)) : 0;
+  const txsPerSecond = spanSec > 0 ? tradeCount / spanSec : 0;
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/wash-forensics.ts:55]: Fixed: the burst test's timestamp window was wrong (12 trades over a 1s span = 12 tps, not the documented 6) — corrected to a real 2s span (6 tps) and added a boundary case documenting the >= threshold behavior.

File: engine/program.ts [hunk 9]
```diff
@@ -6559,6 +6601,41 @@ export const program = Effect.gen(function* () {
...
+          // one Helius call per admitted pool — null (fetch failed, switch
+          // off, no sample) fails open and never blocks. The score is
+          // deliberately hard-only at the extreme tail: a real launch's early
+          // volume is naturally concentrated.
+          const washEvidence = washEvidenceByPool.get(poolAddress);
+          if (
+            !enterGateRejected &&
+            config.launchWashForensicsEnabled === true &&
+            washEvidence !== undefined &&
+            washEvidence.suspicious
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/program.ts]: Fixed: the gate now requires launchScanEnabled && launchExecutionEnabled && launchScanPools.has(poolAddress) — a watchlist/market/autonomous pool entering through its own lane is never wash-rejected just for appearing in the launch top-K.

File: engine/program.ts [hunk 10]
```diff
@@ -6559,6 +6601,41 @@ export const program = Effect.gen(function* () {
...
+          // (few wallets producing the whole recent sample at bot speed)
+          // rejects before capital enters a honeypot's volume. Advisory by
+          // default (switch off); the evidence comes from the radar refresh's
+          // one Helius call per admitted pool — null (fetch failed, switch
+          // off, no sample) fails open and never blocks. The score is
+          // deliberately hard-only at the extreme tail: a real launch's early
+          // volume is naturally concentrated.
+          const washEvidence = washEvidenceByPool.get(poolAddress);
+          if (
+            !enterGateRejected &&
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/program.ts]: Fixed: the wash gate is relocated to the TOP of the ENTER evaluation, before the fallen-angel branch — a suspicious launch pool can no longer route into the FA lane and bypass the evidence check.

File: engine/adapter-service.ts [hunk 11]
```diff
@@ -2104,6 +2105,54 @@ export const AdapterLive = Layer.effect(
...
+            },
+            catch: () => null,
+          }).pipe(Effect.catch(() => Effect.succeed(null)));
+          if (!Array.isArray(parsed) || parsed.length === 0) return null;
+          const rows: WashTradeRow[] = [];
+          for (const tx of parsed) {
+            if (!isObject(tx)) continue;
+            const payer = tx["feePayer"];
+            const timestamp = tx["timestamp"];
+            const fee = tx["fee"];
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/adapter-service.ts:2161]: Fixed: only SUCCESSFUL METEORA txs count as volume — failed txs (err set) and system transfers are excluded, so a single active LP's maintenance txs can no longer satisfy the concentration thresholds. DLMM swaps and LP ops are both type UNKNOWN (Helius has no DLMM model); the residual noise is bounded by the extreme-tail thresholds and the advisory-default switch.

File: engine/program.ts [hunk 12]
```diff
@@ -4211,23 +4216,60 @@ export const program = Effect.gen(function* () {
...
+        // otherwise an old suspicious flag would gate ENTER forever.
+        washEvidenceByPool.clear();
+        const topK = gateResult.ranked.slice(0, config.launchScanTopK ?? 30);
+        const evidences = yield* Effect.forEach(
+          topK,
+          (r) =>
+            adapter.getPoolWashEvidence!(r.pool.address).pipe(
+              Effect.catch(() => Effect.succeed(null)),
+            ),
+          { concurrency: "unbounded" },
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/program.ts]: Fixed: the evidence fetch is bounded in WIDTH (top-30 by fee yield — the pools most likely to reach ENTER; the rest fail open) and CONCURRENCY (5). 200 pools at unbounded concurrency would burst the Helius rate tier and 429 everything to null, silently disabling the gate.

File: engine/adapter-service.ts [hunk 13]
```diff
@@ -2104,6 +2105,63 @@ export const AdapterLive = Layer.effect(
...
+            host = new URL(config.solanaRpcUrl).host;
+          } catch {
+            return null;
+          }
+          if (host !== "helius-rpc.com" && !host.endsWith(".helius-rpc.com")) {
+            return null;
+          }
+          const url =
+            `https://${host}/v0/addresses/${poolAddress}/transactions` +
+            `?limit=40&api-key=${encodeURIComponent(config.heliusApiKey)}`;
```
[the remark this answers is not in the document]
Comment (irfndi) [reply engine/adapter-service.ts:2137]: Fixed: the adapter now maps the RPC host to the enhanced API host (mainnet.helius-rpc.com -> api-mainnet.helius-rpc.com, devnet likewise — verified HTTP 200 with data; the bare helius-rpc.com host is kept). Reusing the RPC host silently nulled every response under the standard setup, so the gate never fired.
